### ISSUE TITLE
macOS launch script shebang

### ISSUE THREAD

**Issue by carlosperate** [mu-editor/pup#153]
So my original idea was based on this python script working as  the .app package launcher (`mu-editor.app/Contents/MacOS/Mu Editor`) when the shebang full path is replaced with the read app bundle path:
```python
#!/full/path/to/mu-editor.app/Contents/Resources/Python/bin/python3.8

# -*- coding: utf-8 -*-
from mu.app import run


if __name__ == "__main__":
    run()

```

With this launcher script I was able to flash the micro:bit without giving shell disk access. 🎉 

However, the problem in this case is having a shebang that can launch a Python interpreter with a relative path (`../Resources/Python/bin/python3.8`). I remembered I saw something similar in the virtualenv project, but they changed it to this, which uses a shell so it doesn't help us :(

```
#!/bin/sh
'''exec' "$(dirname "$0")/../Resources/Python/bin/python3.8" "$0" "$@"
' '''
```

**Comment by tmontes** [comment]
Thanks for the note @carlosperate. Here are a few thoughts:

* Absolute path to the bundled Python binary is not workable, as you mention.
* WRT a relative-path shebang I'm linking a SO article but, I suppose, that might lead us to the same place: something other than "our app" needs to be given shell access (be it awk, perl, or whatever).
* Another crazy-but-might-work approach:
  * Replace the current shell script with a copy of the Python executable (or hard link? or move it over?). This might make it difficult for Python to automatically determine `sys.path` (needs investigation).
  * Have the `Info.plist` call the app binary with the `-m <launch-module>` args (no idea if possible, but maybe!).

**Comment by tmontes** [comment]
SELF-NOTE:

* Hacked this solution away, the other day.
* Bringing it here, just in case.
* Idea:
  * Have a C-written shim that will be the app bundle's executable: gets file paths straight, sets env vars as needed, and `exec`s the correct `.../python -m <launch-module`.
  * `pup` will get it with the macOS template layout, compile it, and remove the C source file.

ALSO:
* In this example, the PBS `python` binary is hard-renamed to `<nice-name>` in order to have macOS show <nice-name> it its menu.

```
#include <libgen.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>

#include <mach-o/dyld.h>


#define PATH_BUFFER_SIZE 4096


int main() {

    char * nice_name = "PuPPY";

    char launcher_path[PATH_BUFFER_SIZE];
    char launcher_dir[PATH_BUFFER_SIZE];
    char python_path[PATH_BUFFER_SIZE];
    char tcl_lib_path[PATH_BUFFER_SIZE];
    char tcl_lib_env[PATH_BUFFER_SIZE];
    uint32_t size = PATH_BUFFER_SIZE;

    if ( _NSGetExecutablePath(launcher_path, &size) != 0 ) {
        fprintf(stderr, "_NSGetExecutablePath failed: need %u bytes buffer.\n", size);
    }
    strncpy(launcher_dir, dirname(launcher_path), PATH_BUFFER_SIZE);
    int i = strlen(launcher_dir)-1;
    if ( launcher_dir[i] == '/' ) {
        launcher_dir[i] = 0;
    }

    snprintf(tcl_lib_path, PATH_BUFFER_SIZE, "%s/../Resources/Python/lib/tcl8.6", launcher_dir);
    snprintf(tcl_lib_env, PATH_BUFFER_SIZE, "TCL_LIBRARY=%s", tcl_lib_path);
    putenv(tcl_lib_env);

    snprintf(python_path, PATH_BUFFER_SIZE, "%s/../Resources/Python/bin/%s", launcher_dir, nice_name);
    char * args[] = {nice_name, "-m" , "puppy"};
    if ( execv(python_path, args) ) {
        fprintf(stderr, "execv(\"%s\") failed: %s.\n", python_path, strerror(errno));
    }

    return 0;
}
```

**Comment by tmontes** [comment]
Closing after having merged #173.
